contracts/contracts/vault/OETHVaultCore.sol

```diff
@@ -345,23 +384,79 @@ contract OETHVaultCore is VaultCore {
+            if (balance + queue.claimed >= queue.queued) {
+                return balance + queue.claimed - queue.queued;
+            }
         }
     }
 
-    function _allocate() internal override {
+    /**
+     * @notice Allocate unallocated funds on Vault to strategies.
+     **/
+    function allocate() external override whenNotCapitalPaused nonReentrant {
         // Add any unallocated WETH to the withdrawal queue first
         _addWithdrawalQueueLiquidity();
 
-        super._allocate();
+        _allocate();
     }
 
+    /// @dev Allocate WETH to the default WETH strategy if there is excess to the Vault buffer.
+    /// This is called from either `mint` or `allocate` and assumes `_addWithdrawalQueueLiquidity`
+    /// has been called before this function.
+    function _allocate() internal override {
+        // No need to do anything if no default strategy for WETH
+        address depositStrategyAddr = assetDefaultStrategies[weth];
+        if (depositStrategyAddr == address(0)) return;
+
+        uint256 wethAvailableInVault = _wethAvailable();
+        // No need to do anything if there isn't any WETH in the vault to allocate
+        if (wethAvailableInVault == 0) return;
+
+        // Calculate the target buffer for the vault using the total supply
+        uint256 totalSupply = oUSD.totalSupply();
+        uint256 targetBuffer = totalSupply.mulTruncate(vaultBuffer);
+
+        // If available WETH in the Vault is below or equal the target buffer then there's nothing to allocate
+        if (wethAvailableInVault <= targetBuffer) return;
+
+        // The amount of assets to allocate to the default strategy
+        uint256 allocateAmount = wethAvailableInVault - targetBuffer;
+
+        IStrategy strategy = IStrategy(depositStrategyAddr);
+        // Transfer WETH to the strategy and call the strategy's deposit function
+        IERC20(weth).safeTransfer(address(strategy), allocateAmount);
+        strategy.deposit(weth, allocateAmount);
+
+        emit AssetAllocated(weth, depositStrategyAddr, allocateAmount);
+    }
+
+    /// @dev The total value of all assets held by the vault and all its strategies
+    /// less any WETH that is reserved for the withdrawal queue.
+    /// For OETH, this is just WETH in the vault and strategies.
+    ///
+    // If there is not enough WETH in the vault and all strategies to cover all outstanding
+    // withdrawal requests then return a total value of 0.
     function _totalValue() internal view override returns (uint256 value) {
-        value = super._totalValue();
+        value = _totalValueInVault() + _totalValueInStrategies();
 
         // Need to remove WETH that is reserved for the withdrawal queue.
         // reserved for the withdrawal queue = cumulative queued total - total claimed
         WithdrawalQueueMetadata memory queue = withdrawalQueueMetadata;
-        value = value + queue.claimed - queue.queued;
+        if (value + queue.claimed >= queue.queued) {
```

Comment:
🟥 This appears to be extremely wrong.

As `queued` increases, the total returned goes down. However when this if statement flips, the total instantly jumps up to the balance, ignoring `claimed` and `queued`

<img width="440" alt="image" src="https://github.com/user-attachments/assets/33090aa0-b778-4b28-845e-1b6ae8c9a89d">

Example:

> Vault controls 100 million WETH.
>
> Someone requests a 50 million withdraw. Vault now reports 50 million total value. Correct.
> Someone requests a 49 million withdraw next. Vault now reports 1 million total value. Correct.
> Someone requests a 1 million + 1 raw withdraw next. Vault now reports 100 million total value instead of zero. Not good.

This could result in an insane rebase, considering at this point there would be a low totalSupply.


This also applies to `_checkBalance`, which uses a similar if statement.

Reply:
Good pickup and an embarrassing bug by me.

This bug can triggered with this test scenario

1. mint 100 WETH
WETH in vault 100
queued 0
claimed 0
vault + claimed < queue (100 + 0 <= 0) is false
Total value = 100

2. Request withdraw for 90
WETH in vault 100
queue 90
claimed 0
vault + claimed <= queue (100 + 0 <= 90) is false
total value 100 when it should be 10

Reply:
I've pushed a fix that handles the scenario where the vault becomes insolvent enough that the total value in the vault and all strategies is less than the outstanding withdrawal requests.

For example, there was a mass slashing event and most users request to withdraw.

Reply:
For purposing of reviewing this bug, here's the problem function
```Solidity
function _totalValue() internal view override returns (uint256 value) {
    value = _totalValueInVault() + _totalValueInStrategies();

    // Need to remove WETH that is reserved for the withdrawal queue.
    // reserved for the withdrawal queue = cumulative queued total - total claimed
    WithdrawalQueueMetadata memory queue = withdrawalQueueMetadata;
    if (value + queue.claimed >= queue.queued) {
        return value + queue.claimed - queue.queued;
    }
}
```

Here's the new implementation
```Solidity
function _totalValue() internal view override returns (uint256 value) {
    value = _totalValueInVault() + _totalValueInStrategies();

    // Need to remove WETH that is reserved for the withdrawal queue.
    WithdrawalQueueMetadata memory queue = withdrawalQueueMetadata;
    // reserved for the withdrawal queue = cumulative queued total - total claimed
    uint256 reservedForQueue = queue.queued - queue.claimed;

    if (value < reservedForQueue) {
        // This can happen if the vault becomes insolvent enough that the
        // total value in the vault and all strategies is less than the outstanding withdrawals.
        // For example, there was a mass slashing event and most users request a withdrawal.
        return 0;
    }

    // Adjust the total value by the amount reserved for the withdrawal queue
    return value - reservedForQueue;
}
```